apps/tailwind-components/components/input/Listbox.vue

```diff
@@ -0,0 +1,345 @@
+    required?: boolean;
+    hasError?: boolean;
+    placeholder?: string;
+    disabled?: boolean;
+  }>(),
+  {
+    disabled: false,
+    required: false,
+    hasError: false,
+    placeholder: "Select an option",
+  }
+);
+
+const sourceDataType = ref<string>("");
+const sourceData = ref<IListboxOption[]>();
+const focusCounter = ref<number>(0);
+const isExpanded = ref<boolean>(false);
+const modelValue = defineModel<IListboxOption | IListboxValueArray>();
+const ulElemRef = useTemplateRef<HTMLUListElement>("listbox-ul");
+const olElemRefs = useTemplateRef<HTMLOptionElement[]>("listbox-li");
+const btnElemRef = useTemplateRef<HTMLButtonElement>("listbox-button");
+const listboxDisplayText = ref<string>(props.placeholder);
+const startingCounter = ref<number>(0);
+
+function counterIsInRange(value: number) {
+  return value <= listboxOptions.value.length - 1 && value >= 0;
+}
+
+const listboxOptions = computed<IInternalListboxOption[]>(() => {
+  const testDataValue: IListboxOption | IListboxValueArray = props.options[0];
+  if (
+    typeof testDataValue === "object" &&
+    Object.hasOwn(testDataValue, "value")
+  ) {
+    sourceDataType.value = "ListboxOptions";
+    sourceData.value = props.options as IListboxOption[];
+  } else {
+    sourceDataType.value = "ListboxValueArray";
+    const inputData = props.options as IListboxValueArray[];
+    const processedData = inputData.map(
+      (value: IListboxValueArray): IListboxOption => {
+        return { value: value };
+      }
+    );
+    sourceData.value = processedData as IListboxOption[];
+  }
+
+  const defaultOption: IListboxOption = {
+    value: null,
+    label: props.placeholder,
+  };
+  const inputData = sourceData.value as IListboxOption[];
+  const data: IListboxOption[] = [defaultOption, ...inputData];
+
+  return data.map((option: IListboxOption, index: number) => {
+    return {
+      ...option,
+      index: index,
+      elemId: `listbox-${props.id}-options-${index}`,
+    };
+  });
+});
+
+function updateCounter(value: number) {
+  if (counterIsInRange(value)) {
+    focusCounter.value = value;
+  } else {
+    if (value > listboxOptions.value.length - 1) {
+      focusCounter.value = listboxOptions.value.length - 1;
+    } else {
+      focusCounter.value = 0;
+    }
+  }
+}
+
+function focusListOption() {
+  nextTick(() => {
+    if (olElemRefs.value) {
+      const targetElem = olElemRefs.value[focusCounter.value];
+      targetElem.setAttribute("tabindex", "0");
+      targetElem.focus();
+    }
+  });
+}
+
+function blurListOption(option: IInternalListboxOption) {
+  nextTick(() => {
+    if (olElemRefs.value) {
+      const targetElem = olElemRefs.value[option.index];
+      targetElem.setAttribute("tabindex", "-1");
+      targetElem.blur();
+    }
+  });
+}
+
+function updateListboxDisplayText(text: string | undefined | null): string {
+  if (typeof text === "undefined") {
+    return props.placeholder;
+  }
+  return text as string;
```

Comment:
Might be nice to instead of doing 'as string' do a 'JSON.stringify(text)' to make it an actual string.

Reply:
Looks like other typing issues were fixed and `as string` is no longer needed.